### ISSUE TITLE
Using Conex as decoder

### ISSUE THREAD

**Issue by Demirrr** [dice-group/Convolutional-Complex-Knowledge-Graph-Embeddings#7]
Coppied from https://github.com/dice-group/Convolutional-Complex-Knowledge-Graph-Embeddings/issues/6#issuecomment-974040246

Hi, I add ConEx in my own model to predict the score as decoder. While Encoder is based on GCN and attention network. But ConEx get lower MRR than InteractE/ConvE in this situation. So I remove my Encoder and I find ConEx is still far from the performance in the paper. It get 0.33 Valid MRR about 200 epoch. I use the the parameters given in the paper:
embedding_dim = 100 ; kernel_size = 3; num_of_output_channels = 16; input_dropout = 0.3; feature_map_dropout = 0.4; batch_size = 512; learning_rate = 0.01; label_smoothing = 0.1 and seed = 1.
Is there something else I haven't noticed? The code all I used is the 'class ConEx' in complex_models.py. I used BCE loss and KvsAll scoring technique.

**Comment by Demirrr** [comment]
Let's concretize the above mentioned issue.

### 1. Using ConEx as a decoder on a GCN encoder
it is difficult to detect the issue as many important details are not provided. It is unclear how a knowledge graph embedding model can be used as a decoder. KGE models often transforms a triple into a real value, i.e., a kge can be considered as encoder. I am not quite sure how you designed your approach. Importantly, getting lower MRR on a test data may stem from (a) overfitting or (b) unoptimized proper training. 


### 2. Suitable hyperparameter configuration

Finding a suitable hyperparameter setting on an input dataset is not a trivial task. If you are using a benchmark dataset for the link prediction, I would recommend you to take a look at our pretrained models. Therein, you can find the applied hyperparameter configuration. Moreover, I am afraid you are mixing something as we did not the above written hyperparameter configuration in our paper. 

Let me know if the above explanation do not help you to solve your problem.

**Comment by Jim-Hou** [comment]
Thank you very much for your reply! I download your pretranined models and use the applied hyperparameter configuration to train. The result is better than before, but still not reach the 0.366 MRR mentioned in paper. I get the number 0.355. Maybe my code has something wrong in training and evaluating process. 
In addition, the use of the GCN encoder I mentioned is to capture structure information. I use the embedding combined with neighbor information to predict the score, instead of initial embedding. However, I actually use initial embedding to reproduce your model result.
Thanks again.

**Comment by Demirrr** [comment]
Hello @Jim-Hou,

## Implementation of the filtered link predicition evaluation
I would suggest you to doublecheck your implementation of the filtered link prediction evaluation, e.g., set a breakpoint and go one by one to ensure the correctness. I am afraid there might be some issue within your implementation of the filtered link prediction evaluation.

## GCN for capturing more information
Incorporating neighborhood information (maybe even neighborhood of neighborhoods) in computing s score of triples is a indeed a good idea. I would expect that this would increase the generalization performance in the link prediction task. I reckon you may find these papers useful [Relational Message Passing for KGC](https://www-cs.stanford.edu/people/jure/pubs/pathcon-kdd21.pdf) and [Inductive Learning on Commonsense KGC](https://arxiv.org/pdf/2009.09263.pdf). Keep me posted about your progress if you wish :)

**Comment by Jim-Hou** [comment]
Thanks !!! :)

**Comment by Jim-Hou** [comment]
Dear **@Demirrr**
Hi, I tried to repuduce your model and I realized that you use both training and validation data to train ConEx.  To be honest, this method puzzles me. For example, It's like my classmates study with the textbook, and occasionally there will be some small tests to verify  them. We imagine textbook and small tests describe training data and validation data. Moreover, test data is the final examination. If not only I learned the knowledge in textbook, but also in small tests, which should have been unknown. I definitely can get a high score in final examination. But I wonder whether it is reasonable.

Besides, I'm not sure the baseline mentioned in your paper were trained on the combination of train and valid sets.
I'm not challenging your research, just to discuss the question :）If this method is completely reasonable, I will use it on my own model.

**Comment by Demirrr** [comment]
Hello @Jim-Hou,

We applied the very same experimental setting described in Section 5.3 to train ConEx and other baselines (DistMult, ComplEx and TuckER).  Hence, I reckon it is indeed reasonable. 

That being said, I suppose quantifying the degree of reasonableness is a nontrival task :). One may argue that if all other remaining models were not retrained on the train & validation sets and their hyperparameters were only tunned with few configurations, then this may indeed favored peformance of ConEx compared to these models. Analogously, one may also argue that it is unreasonable to explore more than 500 unique hyperparameter configurations and/or tunning the seed of random number generator.

Lastly, I humbly do suggest you to challenge other's research including mine :) so that the quality of research in this domain can increase.
